### PR TITLE
feat: add Wally

### PR DESCRIPTION
**Pull request by guidojw**
Adds support for the Wally package manager.
https://github.com/renovatebot/renovate/discussions/25339

~~Installing Wally on Ubuntu focal will cause libssl errors. I tried installing the correct libssl here in a prepare class but that was too much of a hassle where it needs jammy apt sources in order to even be able to find the version, but for its dependencies too, so in the end I made it throw an error instead.~~

### CODE REVIEW

**Comment by guidojw** [comment]
Because of the focal limitation, the test/latest tests here are blocked by https://github.com/containerbase/base/pull/1991.
Shall I comment them or wait for that to be merged?

**Comment by rarkins** [comment]
@guidojw do you mean that wally will only work on Ubuntu 22.04?

**Comment by guidojw** [comment]
> @guidojw do you mean that wally will only work on Ubuntu 22.04?

Yes, Wally only works with libssl3 and focal comes with 1.1.
I tried installing 3 on focal but this meant pointing apt repositories for libssl specifically to the jammy ones, which worked but then it started erroring about libc6 also not being the right version.
I could point libc6 to jammy repositories too but read online that this is not a good idea because lots of os components might depend on it.

**Comment by rarkins** [comment]
Right now we're still defaulting to Ubuntu 20 for our builds I think

**Comment by guidojw** [comment]
Yeah, https://github.com/containerbase/base/pull/1991 should be the only thing blocking this right? I'll convert to draft until that is merged

File: test/latest/Dockerfile [hunk 1]
```diff
@@ -190,7 +190,7 @@ RUN prepare-tool all
 RUN set -ex; [ -d /usr/local/erlang ] && echo "works" || exit 1;
 
 #--------------------------------------
-# test: bazelisk, bun, vendir, helmfile, kustomize
+# test: bazelisk, bun, gleam, helmfile, kustomize, vendir, wally
```
Comment (viceice) [inline test/latest/Dockerfile:193]: please create a seperate refactor PR for the reorderings

File: test/latest/Dockerfile [hunk 2]
```diff
@@ -203,45 +203,56 @@ RUN install-tool bun 1.0.28
...
+ARG WALLY_VERSION=0.3.2
 
 RUN install-tool helmfile "v${HELMFILE_VERSION}"
 
 RUN install-tool kustomize "${KUSTOMIZE_VERSION}"
 
-RUN set -ex; vendir --version
+RUN install-tool vendir "v${VENDIR_VERSION}"
+
+RUN install-tool wally "v${WALLY_VERSION}"
```
Comment (viceice) [inline test/latest/Dockerfile:224]: this will probabbly fail, because latest depends on 20.04

**Review by viceice — COMMENTED** [review]
did you validated that wally uses system libssl and doesn't include it?

**Comment by guidojw** [comment]
I'll resolve the comments and split those changes into another PR soon then 👍🏿

**Comment by viceice** [comment]
https://github.com/containerbase/wally-prebuild/releases/tag/0.3.2 💪  aarch64 binaries are coming in a couple of minutes

**Comment by viceice** [comment]
@guidojw any chance to continue soon?

**Comment by guidojw** [comment]
@viceice resolved comments and updated code to latest project standards 👍🏿

**Comment by guidojw** [comment]
The prebuilds repository README says focal is supported, but the code here will error on that, is that a problem? https://github.com/containerbase/wally-prebuild?tab=readme-ov-file#local-development

File: docs/custom-registries.md [hunk 3]
```diff
@@ -583,3 +583,17 @@ Samples:
 ```txt
 https://github.com/vmware-tanzu/carvel-vendir/releases/download/v0.22.0/vendir-linux-amd64
 ```
+
+## `wally`
+
+Wally releases are downloaded from:
+
+- `https://github.com/UpliftGames/wally/releases`
```
Comment (viceice) [inline docs/custom-registries.md:591]: use our pre built urls only

File: src/cli/tools/wally.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  override async install(version: string): Promise<void> {
+    const distro = await getDistro();
+    // wally requires libssl3 which is not easily installable on focal
+    if (distro.versionCode === 'focal') {
+      throw new Error(`Unsupported distro: ${distro.versionCode}`);
+    }
+
+    const baseUrl = `https://github.com/UpliftGames/wally/releases/download/v${version}/`;
```
Comment (viceice) [inline src/cli/tools/wally.ts:34]: use our own pre built binaries, they work on focal

**Review by viceice — CHANGES_REQUESTED** [review]

File: src/cli/tools/wally.ts (hunk 4 above)
Comment (guidojw) [reply src/cli/tools/wally.ts:34]: Can I remove the distro check in the installer then?

Comment (viceice) [reply src/cli/tools/wally.ts:34]: yes

Comment (guidojw) [reply src/cli/tools/wally.ts]: Done

File: test/latest/Dockerfile [hunk 5]
```diff
@@ -205,7 +205,7 @@ RUN prepare-tool all
 RUN set -ex; [ -d /usr/local/erlang ] && echo "works" || exit 1;
 
 #--------------------------------------
-# test: bazelisk, bun, devbox, vendir, helmfile, kustomize, skopeo
+# test: bazelisk, bun, devbox, vendir, helmfile, kustomize, skopeo, wally
```
Comment (viceice) [inline test/latest/Dockerfile:208]: revert this file, the distro test is enough and does basically the same

**Comment by viceice** [comment]
image
use the re-request review button to request a new review 😉

File: docs/custom-registries.md [hunk 6]
```diff
@@ -583,3 +583,16 @@ Samples:
...
+
+Wally releases are downloaded from:
+
+- `https://github.com/containerbase/wally-prebuild/releases`
+
+Samples:
+
+```txt
+https://github.com/containerbase/wally-prebuild/releases/0.3.2/wally-0.3.2-jammy-x86_x64.tar.xz.sha512
+https://github.com/containerbase/wally-prebuild/releases/0.3.2/wally-0.3.2-jammy-x86_x64.tar.xz
```
Comment (viceice) [inline docs/custom-registries.md:597]:
```suggestion
https://github.com/containerbase/wally-prebuild/releases/download/0.3.2/wally-0.3.2-jammy-x86_x64.tar.xz.sha512
https://github.com/containerbase/wally-prebuild/releases/download/0.3.2/wally-0.3.2-jammy-x86_x64.tar.xz
```

**Review by viceice — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
you need a noble workaround: https://github.com/containerbase/base/blob/9f9475abf07266bf23ba726b54f88ffd803c39e3/src/cli/tools/php/index.ts#L93-L96

**Review by viceice — APPROVED** [review]

**Pull request merged**